File: projects/repo-intake-fbabaei-microsoft-azure-architecture-factory-20260425012922/projects/mdr-support-20260421003634/src/Services/ExtractionService.cs
namespace GeneratedApi.Services;

using System.Text.RegularExpressions;
using GeneratedApi.Models;

public class ExtractionService
{
    // Mandatory domain fields the clarification loop expects.
    public static readonly string[] MandatoryFields =
    {
        "reference_id",
        "submission_date",
        "jurisdiction",
        "summary",
    };

    public ExtractionResult Extract(IngestedDocument doc)
    {
        var text = doc.TextExcerpt ?? string.Empty;
        var fields = new Dictionary<string, string?>();

        var ref_ = Regex.Match(text, @"\b(?:ref|reference)\s*[:#]?\s*([A-Z0-9\-]{4,})", RegexOptions.IgnoreCase);
        if (ref_.Success) fields["reference_id"] = ref_.Groups[1].Value;

        var date = Regex.Match(text, @"\b(20\d{2}-\d{2}-\d{2})\b");
        if (date.Success) fields["submission_date"] = date.Groups[1].Value;

        var jur = Regex.Match(text, @"\bjurisdiction\s*[:=]?\s*([A-Za-z ]{2,40})", RegexOptions.IgnoreCase);
        if (jur.Success) fields["jurisdiction"] = jur.Groups[1].Value.Trim();

        if (text.Length > 0) fields["summary"] = text.Length > 200 ? text[..200] : text;

        return new ExtractionResult(fields, text.Length > 500 ? text[..500] : text);
    }
}

File: projects/repo-intake-fbabaei-microsoft-azure-architecture-factory-20260425012922/projects/mdr-support-20260421002041/src/mdr_support/services/extraction_service.py
"""Structured extraction stub.

Replace ``extract_structured_data`` with an Azure OpenAI / Foundry call
that emits the same ExtractionResult shape in production.
"""
from __future__ import annotations

import re

from ..models import ExtractionResult
from .document_ingestion import IngestedDocument

MANDATORY_FIELDS = ("reference_id", "submission_date", "jurisdiction", "summary")


def extract_structured_data(doc: IngestedDocument) -> ExtractionResult:
    text = doc.text_excerpt or ""
    fields: dict[str, str] = {}
    ref = re.search(r"\b(?:ref|reference)\s*[:#]?\s*([A-Z0-9\-]{4,})", text, re.IGNORECASE)
    if ref:
        fields["reference_id"] = ref.group(1)
    date = re.search(r"\b(20\d{2}-\d{2}-\d{2})\b", text)
    if date:
        fields["submission_date"] = date.group(1)
    jurisdiction = re.search(r"\bjurisdiction\s*[:=]?\s*([A-Za-z ]{2,40})", text, re.IGNORECASE)
    if jurisdiction:
        fields["jurisdiction"] = jurisdiction.group(1).strip()
    if text:
        fields["summary"] = text[:200]
    return ExtractionResult(fields=fields, raw_excerpt=text[:500])
